File: Agent Orchestration Pattern/autogen_typescript_example/agent_orchestration.ts
// Agent Orchestration Pattern - Autogen TypeScript Example
// To run: npm install && npm run agent-orchestration

import axios from 'axios';
import * as readline from 'readline';
import * as dotenv from 'dotenv';
dotenv.config();

const MISTRAL_API_URL = 'https://api.mistral.ai/v1/chat/completions';
const MISTRAL_API_KEY = process.env.MISTRAL_API_KEY;

async function agent(name: string, role: string, input: string): Promise<string> {
  const prompt = `You are ${name}, your role is: ${role}. Here is your input: ${input}`;
  const response = await axios.post(
    MISTRAL_API_URL,
    {
      model: 'mistral-tiny',
      messages: [{ role: 'user', content: prompt }],
    },
    {
      headers: {
        'Authorization': `Bearer ${MISTRAL_API_KEY}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data.choices[0].message.content;
}

async function orchestrator(task: string) {
  // Dynamically assign roles and tasks
  const roles = [
    { name: 'Alice', role: 'Researcher', subtask: 'Find background info about: ' + task },
    { name: 'Bob', role: 'Analyst', subtask: 'Analyze the following info: ' },
    { name: 'Carol', role: 'Presenter', subtask: 'Summarize and present the findings: ' }
  ];

  // Step 1: Researcher
  const research = await agent(roles[0].name, roles[0].role, roles[0].subtask);
  console.log(`${roles[0].name} (${roles[0].role}):`, research);

  // Step 2: Analyst
  const analysis = await agent(roles[1].name, roles[1].role, roles[1].subtask + research);
  console.log(`${roles[1].name} (${roles[1].role}):`, analysis);

  // Step 3: Presenter
  const presentation = await agent(roles[2].name, roles[2].role, roles[2].subtask + analysis);
  console.log(`${roles[2].name} (${roles[2].role}):`, presentation);
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question('Task: ', async (userInput: string) => {
  try {
    await orchestrator(userInput);
  } catch (err) {
    console.error('Error:', err);
  }
  rl.close();
});

File: agent-orchestration-pattern/langgraph_python_example/agent_orchestration.py
# Agent Orchestration Pattern - LangGraph Python Example

This example demonstrates the Agent Orchestration Pattern using LangGraph and Python. An orchestrator coordinates three agents (Researcher, Analyst, Presenter) to solve a task. The LLM is Mistral.

## Requirements

- Python 3.8+
- `langgraph` library
- `python-dotenv` (for .env support)
- Mistral LLM API access

## Install dependencies

```bash
pip install langgraph python-dotenv requests
```

## Example Code

```python
import os
from langgraph import Agent, Environment, LLM
from dotenv import load_dotenv

load_dotenv()

MISTRAL_API_KEY = os.getenv("MISTRAL_API_KEY")
MISTRAL_API_URL = "https://api.mistral.ai/v1/chat/completions"

class SimpleEnvironment(Environment):
    def get_observation(self):
        return input("Task: ")
    def send_action(self, action):
        print(action)

class Researcher(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, task):
        prompt = f"You are Alice, a Researcher. Find background info about: {task}"
        return self.llm.complete(prompt)

class Analyst(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, info):
        prompt = f"You are Bob, an Analyst. Analyze the following info: {info}"
        return self.llm.complete(prompt)

class Presenter(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, analysis):
        prompt = f"You are Carol, a Presenter. Summarize and present the findings: {analysis}"
        return self.llm.complete(prompt)

llm = LLM(
    provider="mistral",
    api_key=MISTRAL_API_KEY,
    api_url=MISTRAL_API_URL,
)

env = SimpleEnvironment()
researcher = Researcher(llm)
analyst = Analyst(llm)
presenter = Presenter(llm)

task = env.get_observation()
research = researcher.act(task)
print("Alice (Researcher):", research)
analysis = analyst.act(research)
print("Bob (Analyst):", analysis)
presentation = presenter.act(analysis)
print("Carol (Presenter):", presentation)
```

---

- Try a research or analysis task to see orchestration in action.
- Make sure your `.env` file contains your Mistral API key.
